pydantic/types.py

```diff
@@ -107,6 +121,60 @@ def validate(cls, value):
         return validate_email(value)[1]
 
 
+class UrlStr(str):
+    strip_whitespace = True
+    min_length = 1
+    max_length = 2 ** 16
+    default_schemes = {'http', 'https', 'ftp', 'ftps'}
+    schemes: Optional[Set[str]] = None
+    relative = False  # whether to allow relative URLs
+    require_tld = True  # whether to reject non-FQDN hostnames
+
+    @classmethod
+    def get_validators(cls):
+        yield not_none_validator
+        yield str_validator
+        yield anystr_strip_whitespace
+        yield anystr_length_validator
+        yield cls.validate
+
+    @classmethod
+    def validate(cls, value: str) -> str:
+        # Check first if the scheme is valid
+        schemes = cls.schemes or cls.default_schemes
+        if '://' in value:
+            scheme = value.split('://')[0].lower()
+            if scheme not in schemes:
+                raise errors.UrlSchemeError(scheme=scheme)
+
+        regex = url_regex_generator(relative=cls.relative, require_tld=cls.require_tld)
```

Comment:
can we generate this on `__init__` then we won't need the lru cache?

Reply:
to my mind better to use `lru_cache` in this case, it can help us to decrease number of calls if `re.compile`